### ISSUE TITLE
Some CLP(X) for Tau Prolog? Is there dif/1 or freeze/1 already?

### ISSUE THREAD

**Issue by jburse** [jariazavalverde/tau-prolog#77]
Hi,

Is there dif/1 or freeze/1 already for Tau Prolog? How would a derivation 
tree visualization look like? I can only contribute in that I have a new
Prolog trampolin design, which could be useful:

**Comment by jariazavalverde** [comment]
Not yet. I'm currently working on the IO builtin predicates for both browser and Node.js (#59). Thanks for the suggestion, I have CLP on my TODO list.

**Comment by jburse** [comment]
CLP(X) is hard. I have not yet solve the GC Problem. All the unify hooks are not
reclaimed, since I don't know how far the choice points go in relation to
the unify hooks. Mercury Prolog has something very refined here:

https://www.mercurylang.org/information/doc-latest/mercury_ref/Avoiding-redundant-trailing.html#Avoiding-redundant-trailing

It might be the case that SWI-Prolog has also something like that, not sure. Maybe
even somewhere in their nb_setarg/3. Dont remember exactly, had a look
once at their source.
https://www.mercurylang.org/information/doc-latest/mercury_ref/Trailing.html

**Comment by JeanChristopheRohner** [comment]
It might be worthwhile to consider clp(BNR); seems very powerful and works for boolean, integers and reals :-)

**Comment by lf94** [comment]
Has this changed at all? Is CLP in some form possible yet?

**Comment by jariazavalverde** [comment]
No. I've been working on a branch to incorporate attributed variables, but I don't know when I'll get back to it.

**Comment by lf94** [comment]
Thank you for your reply! I'm using tau-prolog in my project https://github.com/lf94/paizo-prolog but I've realized now it'd be nice to have this for constraint problems. I'm considering a SWI-Prolog WASM build, but would gladly switch back to tau-prolog. It's an excellent project!

**Comment by gabrielmontagne** [comment]
Hi, another onlooker is coming to check for any progress on this.  IMO this would be a game-changer, tbh, for web development.
